File: sdk/privacy/src/encryption.rs
//! Note Encryption
//!
//! Encrypts note data for the recipient using ECDH + ChaCha20-Poly1305.
//!
//! ```text
//! Flow:
//! 1. Sender generates ephemeral keypair (epk, esk)
//! 2. Shared secret = ECDH(esk, recipient_pk)
//! 3. Encryption key = HKDF(shared_secret, "zelana-note-v1")
//! 4. Ciphertext = ChaCha20-Poly1305(key, nonce, plaintext)
//! 5. Output = (epk, nonce, ciphertext, tag)
//! ```

use chacha20poly1305::{
    ChaCha20Poly1305, Nonce,
    aead::{Aead, KeyInit},
};
use serde::{Deserialize, Serialize};
use x25519_dalek::{EphemeralSecret, PublicKey, StaticSecret};

use crate::note::Note;

/// An encrypted note (sent on-chain)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EncryptedNote {
    /// Ephemeral public key for ECDH
    pub ephemeral_pk: [u8; 32],
    /// Nonce for ChaCha20-Poly1305
    pub nonce: [u8; 12],
    /// Encrypted note data with authentication tag
    pub ciphertext: Vec<u8>,
}

/// Note plaintext format for encryption
#[derive(Debug, Clone, Serialize, Deserialize)]
struct NotePlaintext {
    /// Note value
    value: u64,
    /// Randomness (blinding factor)
    randomness: [u8; 32],
    /// Optional memo (up to 512 bytes)
    memo: Vec<u8>,
}

impl EncryptedNote {
    /// Size of encrypted note (fixed overhead)
    pub const OVERHEAD: usize = 32 + 12 + 16; // epk + nonce + tag

    /// Get the ephemeral public key
    pub fn ephemeral_public_key(&self) -> &[u8; 32] {
        &self.ephemeral_pk
    }
}

/// Encrypt a note for a recipient
///
/// # Arguments
/// * `note` - The note to encrypt
/// * `recipient_pk` - Recipient's X25519 public key
/// * `memo` - Optional memo (max 512 bytes)
///
/// # Returns
/// Encrypted note that can be published on-chain
pub fn encrypt_note(note: &Note, recipient_pk: &[u8; 32], memo: Option<&[u8]>) -> EncryptedNote {
    // Generate ephemeral keypair using thread_rng which is compatible with x25519-dalek
    let mut rng = rand::thread_rng();
    let ephemeral_secret = EphemeralSecret::random_from_rng(&mut rng);
    let ephemeral_pk = PublicKey::from(&ephemeral_secret);

    // ECDH shared secret
    let recipient_key = PublicKey::from(*recipient_pk);
    let shared_secret = ephemeral_secret.diffie_hellman(&recipient_key);

    // Derive encryption key using HKDF
    let encryption_key = derive_note_key(shared_secret.as_bytes(), ephemeral_pk.as_bytes());

    // Create plaintext
    let plaintext = NotePlaintext {
        value: note.value.0,
        randomness: note.randomness,
        memo: memo
            .map(|m| m[..m.len().min(512)].to_vec())
            .unwrap_or_default(),
    };

    let plaintext_bytes = serialize_plaintext(&plaintext);

    // Generate random nonce
    let mut nonce_bytes = [0u8; 12];
    use rand::RngCore;
    rng.fill_bytes(&mut nonce_bytes);
    let nonce = Nonce::from_slice(&nonce_bytes);

    // Encrypt with ChaCha20-Poly1305
    let cipher = ChaCha20Poly1305::new_from_slice(&encryption_key).expect("valid key length");

    let ciphertext = cipher
        .encrypt(nonce, plaintext_bytes.as_slice())
        .expect("encryption should not fail");

    EncryptedNote {
        ephemeral_pk: *ephemeral_pk.as_bytes(),
        nonce: nonce_bytes,
        ciphertext,
    }
}

/// Decrypt a note using recipient's secret key
///
/// # Arguments
/// * `encrypted` - The encrypted note
/// * `recipient_sk` - Recipient's X25519 secret key
/// * `expected_owner_pk` - The expected owner public key (for note reconstruction)
///
/// # Returns
/// The decrypted note, or None if decryption fails
pub fn decrypt_note(
    encrypted: &EncryptedNote,
    recipient_sk: &[u8; 32],
    expected_owner_pk: [u8; 32],
) -> Option<(Note, Vec<u8>)> {
    // Reconstruct shared secret
    let secret = StaticSecret::from(*recipient_sk);
    let ephemeral_pk = PublicKey::from(encrypted.ephemeral_pk);
    let shared_secret = secret.diffie_hellman(&ephemeral_pk);

    // Derive decryption key
    let decryption_key = derive_note_key(shared_secret.as_bytes(), &encrypted.ephemeral_pk);

    // Decrypt
    let cipher = ChaCha20Poly1305::new_from_slice(&decryption_key).ok()?;
    let nonce = Nonce::from_slice(&encrypted.nonce);

    let plaintext_bytes = cipher
        .decrypt(nonce, encrypted.ciphertext.as_slice())
        .ok()?;

    // Deserialize
    let plaintext = deserialize_plaintext(&plaintext_bytes)?;

    // Reconstruct note
    let note = Note::with_randomness(plaintext.value, expected_owner_pk, plaintext.randomness);

    Some((note, plaintext.memo))
}

/// Try to decrypt a note (scan mode - for wallet scanning)
///
/// Returns the note if decryption succeeds and commitment matches
pub fn try_decrypt_note(
    encrypted: &EncryptedNote,
    recipient_sk: &[u8; 32],
    expected_owner_pk: [u8; 32],
    expected_commitment: &[u8; 32],
) -> Option<(Note, Vec<u8>)> {
    let (note, memo) = decrypt_note(encrypted, recipient_sk, expected_owner_pk)?;

    // Verify commitment matches
    let computed_commitment = note.commitment();
    if computed_commitment.as_bytes() == expected_commitment {
        Some((note, memo))
    } else {
        None
    }
}

/// Derive encryption key from shared secret
fn derive_note_key(shared_secret: &[u8], ephemeral_pk: &[u8]) -> [u8; 32] {
    // HKDF using blake3
    let mut hasher = blake3::Hasher::new_derive_key("zelana-note-v1");
    hasher.update(shared_secret);
    hasher.update(ephemeral_pk);
    *hasher.finalize().as_bytes()
}

/// Serialize plaintext for encryption
fn serialize_plaintext(pt: &NotePlaintext) -> Vec<u8> {
    let mut bytes = Vec::with_capacity(8 + 32 + 2 + pt.memo.len());

    // Value (8 bytes, little-endian)
    bytes.extend_from_slice(&pt.value.to_le_bytes());

    // Randomness (32 bytes)
    bytes.extend_from_slice(&pt.randomness);

    // Memo length (2 bytes) + memo
    let memo_len = pt.memo.len() as u16;
    bytes.extend_from_slice(&memo_len.to_le_bytes());
    bytes.extend_from_slice(&pt.memo);

    bytes
}

/// Deserialize plaintext after decryption
fn deserialize_plaintext(bytes: &[u8]) -> Option<NotePlaintext> {
    if bytes.len() < 42 {
        return None; // 8 + 32 + 2 minimum
    }

    let value = u64::from_le_bytes(bytes[0..8].try_into().ok()?);
    let randomness: [u8; 32] = bytes[8..40].try_into().ok()?;
    let memo_len = u16::from_le_bytes(bytes[40..42].try_into().ok()?) as usize;

    if bytes.len() < 42 + memo_len {
        return None;
    }

    let memo = bytes[42..42 + memo_len].to_vec();

    Some(NotePlaintext {
        value,
        randomness,
        memo,
    })
}

/// Batch encryption for multiple outputs
pub fn encrypt_notes(
    notes: &[(Note, [u8; 32], Option<Vec<u8>>)], // (note, recipient_pk, memo)
) -> Vec<EncryptedNote> {
    notes
        .iter()
        .map(|(note, recipient_pk, memo)| encrypt_note(note, recipient_pk, memo.as_deref()))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn generate_keypair() -> ([u8; 32], [u8; 32]) {
        let mut rng = rand::thread_rng();
        let secret = StaticSecret::random_from_rng(&mut rng);
        let public = PublicKey::from(&secret);
        (*secret.as_bytes(), *public.as_bytes())
    }

    #[test]
    fn test_encrypt_decrypt_note() {
        let (recipient_sk, recipient_pk) = generate_keypair();

        let note = Note::with_randomness(1000, recipient_pk, [42u8; 32]);
        let memo = b"test memo";

        let encrypted = encrypt_note(&note, &recipient_pk, Some(memo));
        let (decrypted, decrypted_memo) = decrypt_note(&encrypted, &recipient_sk, recipient_pk)
            .expect("decryption should succeed");

        assert_eq!(decrypted.value.0, note.value.0);
        assert_eq!(decrypted.randomness, note.randomness);
        assert_eq!(decrypted_memo, memo);
    }

    #[test]
    fn test_wrong_key_fails() {
        let (_, recipient_pk) = generate_keypair();
        let (wrong_sk, _) = generate_keypair();

        let note = Note::with_randomness(1000, recipient_pk, [42u8; 32]);
        let encrypted = encrypt_note(&note, &recipient_pk, None);

        let result = decrypt_note(&encrypted, &wrong_sk, recipient_pk);
        assert!(result.is_none(), "wrong key should fail decryption");
    }

    #[test]
    fn test_commitment_verification() {
        let (recipient_sk, recipient_pk) = generate_keypair();

        let note = Note::with_randomness(1000, recipient_pk, [42u8; 32]);
        let commitment = note.commitment();

        let encrypted = encrypt_note(&note, &recipient_pk, None);

        // Should succeed with correct commitment
        let result = try_decrypt_note(
            &encrypted,
            &recipient_sk,
            recipient_pk,
            commitment.as_bytes(),
        );
        assert!(result.is_some());

        // Should fail with wrong commitment
        let wrong_commitment = [0u8; 32];
        let result = try_decrypt_note(&encrypted, &recipient_sk, recipient_pk, &wrong_commitment);
        assert!(result.is_none());
    }
}

File: sdk/typescript/src/encryption.ts
/**
 * Note Encryption
 *
 * Encrypts note data for the recipient using ECDH + ChaCha20-Poly1305.
 * This implementation matches the Rust privacy SDK exactly.
 *
 * Flow:
 * 1. Sender generates ephemeral X25519 keypair (epk, esk)
 * 2. Shared secret = ECDH(esk, recipient_pk)
 * 3. Encryption key = HKDF(shared_secret, "zelana-note-v1")
 * 4. Ciphertext = ChaCha20-Poly1305(key, nonce, plaintext)
 * 5. Output = (epk, nonce, ciphertext)
 */

import { x25519 } from '@noble/curves/ed25519.js';
import { chacha20poly1305 } from '@noble/ciphers/chacha.js';
import { blake3 } from '@noble/hashes/blake3';
import { randomBytes as nobleRandomBytes } from '@noble/hashes/utils';
import type { Bytes32 } from './types.js';

/**
 * Encrypted note structure (matches Rust EncryptedNote)
 */
export interface EncryptedNote {
  /** Ephemeral public key for ECDH (32 bytes) */
  ephemeralPk: Uint8Array;
  /** Nonce for ChaCha20-Poly1305 (12 bytes) */
  nonce: Uint8Array;
  /** Encrypted note data with authentication tag */
  ciphertext: Uint8Array;
}

/**
 * Note plaintext structure
 */
interface NotePlaintext {
  /** Note value in lamports */
  value: bigint;
  /** Random blinding factor (32 bytes) */
  randomness: Uint8Array;
  /** Optional memo (max 512 bytes) */
  memo?: Uint8Array;
}

/**
 * Derive encryption key from shared secret using BLAKE3
 * Matches Rust: blake3::Hasher::new_derive_key("zelana-note-v1")
 */
function deriveNoteKey(sharedSecret: Uint8Array, ephemeralPk: Uint8Array): Uint8Array {
  // BLAKE3 key derivation with domain separator
  const context = new TextEncoder().encode('zelana-note-v1');
  
  // Create a keyed hash using the context as a domain separator
  // BLAKE3's derive_key mode is: BLAKE3(context || 0x01 || input)
  // We use a simpler approach that matches the Rust behavior:
  // derive_key(context, input) where input = shared_secret || ephemeral_pk
  const input = new Uint8Array(sharedSecret.length + ephemeralPk.length);
  input.set(sharedSecret, 0);
  input.set(ephemeralPk, sharedSecret.length);
  
  // Use BLAKE3 derive_key mode
  return blake3(input, { context });
}

/**
 * Serialize plaintext for encryption
 * Format: value (8 bytes LE) + randomness (32 bytes) + memo_len (2 bytes LE) + memo
 */
function serializePlaintext(pt: NotePlaintext): Uint8Array {
  const memoLen = pt.memo?.length ?? 0;
  const bytes = new Uint8Array(8 + 32 + 2 + memoLen);
  
  // Value (8 bytes, little-endian)
  const view = new DataView(bytes.buffer);
  view.setBigUint64(0, pt.value, true);
  
  // Randomness (32 bytes)
  bytes.set(pt.randomness, 8);
  
  // Memo length (2 bytes, little-endian)
  view.setUint16(40, memoLen, true);
  
  // Memo
  if (pt.memo && memoLen > 0) {
    bytes.set(pt.memo.slice(0, 512), 42);
  }
  
  return bytes;
}

/**
 * Deserialize plaintext after decryption
 */
function deserializePlaintext(bytes: Uint8Array): NotePlaintext | null {
  if (bytes.length < 42) {
    return null; // 8 + 32 + 2 minimum
  }
  
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  
  // Value (8 bytes, little-endian)
  const value = view.getBigUint64(0, true);
  
  // Randomness (32 bytes)
  const randomness = bytes.slice(8, 40);
  
  // Memo length (2 bytes)
  const memoLen = view.getUint16(40, true);
  
  if (bytes.length < 42 + memoLen) {
    return null;
  }
  
  // Memo
  const memo = memoLen > 0 ? bytes.slice(42, 42 + memoLen) : undefined;
  
  return { value, randomness, memo };
}

/**
 * Encrypt a note for a recipient
 *
 * @param value - Note value in lamports
 * @param randomness - Random blinding factor (32 bytes)
 * @param recipientPk - Recipient's X25519 public key
 * @param memo - Optional memo (max 512 bytes)
 * @returns Encrypted note
 */
export function encryptNote(
  value: bigint,
  randomness: Bytes32,
  recipientPk: Bytes32,
  memo?: Uint8Array
): EncryptedNote {
  // Generate ephemeral keypair
  const ephemeralSk = nobleRandomBytes(32);
  const ephemeralPk = x25519.getPublicKey(ephemeralSk);
  
  // ECDH shared secret
  const sharedSecret = x25519.getSharedSecret(ephemeralSk, recipientPk);
  
  // Derive encryption key
  const encryptionKey = deriveNoteKey(sharedSecret, ephemeralPk);
  
  // Create plaintext
  const plaintext = serializePlaintext({
    value,
    randomness: new Uint8Array(randomness),
    memo: memo?.slice(0, 512),
  });
  
  // Generate random nonce (12 bytes)
  const nonce = nobleRandomBytes(12);
  
  // Encrypt with ChaCha20-Poly1305
  const cipher = chacha20poly1305(encryptionKey, nonce);
  const ciphertext = cipher.encrypt(plaintext);
  
  return {
    ephemeralPk,
    nonce,
    ciphertext,
  };
}

/**
 * Decrypt a note using recipient's secret key
 *
 * @param encrypted - The encrypted note
 * @param recipientSk - Recipient's X25519 secret key (32 bytes)
 * @returns Decrypted plaintext or null if decryption fails
 */
export function decryptNote(
  encrypted: EncryptedNote,
  recipientSk: Bytes32
): { value: bigint; randomness: Uint8Array; memo?: Uint8Array } | null {
  try {
    // ECDH shared secret
    const sharedSecret = x25519.getSharedSecret(recipientSk, encrypted.ephemeralPk);
    
    // Derive encryption key
    const encryptionKey = deriveNoteKey(sharedSecret, encrypted.ephemeralPk);
    
    // Decrypt with ChaCha20-Poly1305
    const cipher = chacha20poly1305(encryptionKey, encrypted.nonce);
    const plaintext = cipher.decrypt(encrypted.ciphertext);
    
    // Deserialize
    const pt = deserializePlaintext(plaintext);
    if (!pt) {
      return null;
    }
    
    return {
      value: pt.value,
      randomness: pt.randomness,
      memo: pt.memo,
    };
  } catch {
    // Decryption failed (wrong key, tampered ciphertext, etc.)
    return null;
  }
}

/**
 * Serialize an encrypted note to bytes for transmission
 * Format: ephemeral_pk (32) + nonce (12) + ciphertext (variable)
 */
export function serializeEncryptedNote(note: EncryptedNote): Uint8Array {
  const bytes = new Uint8Array(32 + 12 + note.ciphertext.length);
  bytes.set(note.ephemeralPk, 0);
  bytes.set(note.nonce, 32);
  bytes.set(note.ciphertext, 44);
  return bytes;
}

/**
 * Deserialize an encrypted note from bytes
 */
export function deserializeEncryptedNote(bytes: Uint8Array): EncryptedNote | null {
  if (bytes.length < 44 + 42 + 16) {
    // Minimum: 32 (epk) + 12 (nonce) + 42 (min plaintext) + 16 (tag)
    return null;
  }
  
  return {
    ephemeralPk: bytes.slice(0, 32),
    nonce: bytes.slice(32, 44),
    ciphertext: bytes.slice(44),
  };
}

/**
 * Generate an X25519 keypair
 */
export function generateX25519Keypair(): { secretKey: Uint8Array; publicKey: Uint8Array } {
  const secretKey = nobleRandomBytes(32);
  const publicKey = x25519.getPublicKey(secretKey);
  return { secretKey, publicKey };
}

/**
 * Derive X25519 public key from secret key
 */
export function x25519PublicKey(secretKey: Bytes32): Uint8Array {
  return x25519.getPublicKey(secretKey);
}
